src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -3016,6 +3033,122 @@ - (BOOL)_attributeAffectsDeviceConfiguration:(MTRAttributePath *)attributePath
     return NO;
 }
 
+- (void)_removeClusters:(NSSet *)clusterPathsToRemove
+{
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    [_persistedClusters minusSet:clusterPathsToRemove];
+
+    for (MTRClusterPath * path in clusterPathsToRemove) {
+        [_persistedClusterData removeObjectForKey:path];
+        [_persistedClusters removeObject:path];
+        [_clusterDataToPersist removeObjectForKey:path];
+        [self.deviceController.controllerDataStore clearStoredClusterDataForNodeID:self.nodeID endpointID:path.endpoint clusterID:path.cluster];
+    }
+}
+
+- (void)_removeAttributes:(NSSet *)toBeRemovedAttributes fromCluster:(MTRClusterPath *)clusterPathToRemoveAttributesFrom
+{
+    if (toBeRemovedAttributes == nil || clusterPathToRemoveAttributesFrom == nil) {
+        return;
+    }
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    for (NSNumber * attribute in toBeRemovedAttributes) {
+        [self _removeCachedAttribute:attribute fromCluster:clusterPathToRemoveAttributesFrom];
+    }
+    [_persistedClusterData removeObjectForKey:clusterPathToRemoveAttributesFrom];
+    [self.deviceController.controllerDataStore clearStoredClusterDataForNodeID:self.nodeID endpointID:clusterPathToRemoveAttributesFrom.endpoint clusterID:clusterPathToRemoveAttributesFrom.cluster];
+}
+
+- (void)_pruneOrphanedEndpoints:(NSDictionary *)previousPartsListValue
+              newPartsListValue:(NSDictionary *)newPartsListValue
+{
+    // If the parts list changed and one or more endpoints were removed, remove all the clusters in _persistedClusters, _persistedClusterData and _clusterDataToPersist for all those endpoints.
+    // Also remove it from the data store.
+    NSMutableSet * toBeRemovedEndpoints = [NSMutableSet setWithArray:[self arrayOfNumbersFromAttributeValue:[self _dataValueWithoutDataVersion:previousPartsListValue]]];
```

Comment:
To be honest, there's lots of usage of non strongly typed data and I was not sure whats the right convention. I used to strongly type when i started writing objective C since it made clear to me what the data structure contains and then I got lazy . will keep in mind :D

Reply:
fixed everywhere